Mathlib/Tactic/MoveAdd.lean

```diff
@@ -0,0 +1,448 @@
+The properties that we mention here assume that this is the case.
+
+Thus, `weight L` is a function `α → ℤ` with the following properties:
+* if `(a, true)  ∈ L`, then `weight L a` is strictly negative;
+* if `(a, false) ∈ L`, then `weight L a` is strictly positive;
+* if neither `(a, true)` nor `(a, false)` is in `L`, then `weight L a = 0`.
+
+Moreover, the function `weight L` is strictly monotone increasing on both
+`{a : α | (a, true) ∈ L}` and `{a : α | (a, false) ∈ L}`,
+in the sense that if `a' = (a, true)` and `b' = (b, true)` are in `L`,
+then `a'` appears before `b'` in `L` if and only if `weight L a < weight L b` and
+similarly for the pairs with second coordinate equal to `false`.
+-/
+def weight (L : List (α × Bool)) (a : α) : ℤ :=
+let l := L.length
+match L.find? (Prod.fst · == a) with
+  | some (_, b) => if b then - l + (L.indexOf (a, b) : ℤ) else (L.indexOf (a, b) + 1 : ℤ)
+  | none => 0
+
+/--  `reorderUsing toReorder instructions` produces a reordering of `toReorder : List α`,
+following the requirements imposed by `instructions : List (α × Bool)`.
+
+These are the requirements:
+* elements of `toReorder` that appear with `true` in `instructions appear at the
+  *beginning* of the reordered list, in the order in which they appear in `instructions`;
+* similarly, elements of `toReorder` that appear with `false` in `instructions appear at the
+  *end* of the reordered list, in the order in which they appear in `instructions`;
+* finally, elements of `toReorder` that do not appear in `instructions` appear "in the middle"
+  with the order that they had in `toReorder`.
+
+For example,
+* `reorderUsing [0, 1, 2] [(0, false)] = [1, 2, 0]`,
+* `reorderUsing [0, 1, 2] [(1, true)] = [1, 0, 2]`,
+* `reorderUsing [0, 1, 2] [(1, true), (0, false)] = [1, 2, 0]`.
+-/
+def reorderUsing (toReorder : List α) (instructions : List (α × Bool)) : List α :=
+let uInstructions :=
+  let (as, as?) := instructions.unzip
+  (uniquify as).zip as?
+let uToReorder := (uniquify toReorder).toArray
+let reorder := uToReorder.qsort fun x y =>
+  match uInstructions.find? (Prod.fst · == x), uInstructions.find? (Prod.fst · == y) with
+    | none, none => (uToReorder.getIdx? x).get! ≤ (uToReorder.getIdx? y).get!
+    | _, _ => weight uInstructions x ≤ weight uInstructions y
+(reorder.map Prod.fst).toList
+
+section tactic
+open Mathlib.MoveAdd
+
+#guard (uniquify [ 0,      1,      0,      1,      0,      3] =
+                 [(0, 0), (1, 0), (0, 1), (1, 1), (0, 2), (3, 0)])
+
+#guard
+  (let dat := [(0, true), (1, false), (2, true)]
+   (#[0, 1, 2, 3, 4].qsort (fun x y => (weight dat x) ≤ (weight dat y)) = #[0, 2, 3, 4, 1]))
+
+#guard false = ( reorderUsing [0, 1, 2] [(0, false)] = [1, 2, 0] &&
+                 reorderUsing [0, 1, 2] [(1, true)] = [1, 0, 2] &&
+                 reorderUsing [0, 1, 2] [(1, true), (0, false)] != [1, 2, 0])
+
+#guard reorderUsing [1, 5, 4, 3, 2, 1] [(3, true), (2, false), (1, false)] =
+                        [3, 5, 4, 1, 2, 1]
```

Comment:
I like this way of testing! Did you take it from some other file? I do wonder if it shouldn't be in the `test` folder...

Reply:
I am glad that you like it!

I think that I had it in the `test` folder, but Scott preferred the tests to actually test the tactic, rather than the internals, so I moved it here,

Reply:
Oh, I see that Scott actually wanted the tests to move to the *bottom* of the test file, not to the *other* file! :facepalm:

I'll move these tests at the bottom of the (currently) other file when I'll look at this PR when I am less tired!